### PR TITLE
Update remote_documents table to add dedicated parent_path column

### PR DESCRIPTION
**Pull request by schmidt-sebastian**
The current table design reads all documents from subcollections as we scan a collection path. This makes it impossible to fetch the first N documents by using "Limit N", which we would like to do for the Index backfill.

### CODE REVIEW

**Comment by schmidt-sebastian** [comment]
I am working on performance data but don't have anything for now.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/local/SQLiteRemoteDocumentCache.java [hunk 1]
```diff
@@ -57,15 +58,16 @@ public void add(MutableDocument document, SnapshotVersion readTime) {
...
     MessageLite message = serializer.encodeMaybeDocument(document);
 
     db.execute(
         "INSERT OR REPLACE INTO remote_documents "
-            + "(path, read_time_seconds, read_time_nanos, contents) "
-            + "VALUES (?, ?, ?, ?)",
-        path,
+            + "(collection_path, document_id, read_time_seconds, read_time_nanos, contents) "
+            + "VALUES (?, ?, ?, ?, ?)",
+        collectionForKey(documentKey),
```
Comment (wu-hui) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/local/SQLiteRemoteDocumentCache.java:69]: These are repeated enough that I think we can add them to `DocumentKey`, like `key.collection()` and `key.documentId()` ?

File: firebase-firestore/src/test/java/com/google/firebase/firestore/local/SQLiteSchemaTest.java [hunk 2]
```diff
@@ -571,6 +571,35 @@ public void rewritesCanonicalIds() {
             });
   }
 
+  @Test
+  public void rewritesDocumentKeys() {
```
Comment (wu-hui) [inline firebase-firestore/src/test/java/com/google/firebase/firestore/local/SQLiteSchemaTest.java:575]: I'd like to also have a test where the number of documents to migrate exceeds the batch size, to make sure batching works.

**Review by wu-hui — COMMENTED** [review]
Looks good. 

How was the performance of the migration?

File: firebase-firestore/src/main/java/com/google/firebase/firestore/local/SQLiteRemoteDocumentCache.java (hunk 1 above)
Comment (schmidt-sebastian) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/local/SQLiteRemoteDocumentCache.java]: Done

File: firebase-firestore/src/test/java/com/google/firebase/firestore/local/SQLiteSchemaTest.java (hunk 2 above)
Comment (schmidt-sebastian) [reply firebase-firestore/src/test/java/com/google/firebase/firestore/local/SQLiteSchemaTest.java]: Done

**Comment by schmidt-sebastian** [comment]
I had to update the PR to add a dedicated parent_path column but keep the other column as is. This allows schema downgrades, as older clients expect `path` to exist. The small additional benefit that this brings is that the schema migration is now a lot less performance intensive and should be okay to ship as we have done in the past. I will provide performance data by the end of the day.

**Comment by schmidt-sebastian** [comment]
FYI:

```
Migration from version 12 to 13 took 408 milliseconds
```

This is for 1775 docs.

**Comment by schmidt-sebastian** [comment]
FYI there is a new version of this PR: https://github.com/firebase/firebase-android-sdk/pull/3192/files

**Pull request closed without merging**